pkg/http-tunnel/server/server.go

```diff
@@ -0,0 +1,762 @@
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Server is responsible for proxying public connections to the client over a
+// tunnel connection.
+type Server struct {
+	*registry
+	config *serverData
+
+	listener    net.Listener
+	connPool    *connPool
+	httpClient  *http.Client
+	logger      *log.Entry
+	vhostMuxer  *vhost.TLSMuxer
+	PortPool    *AddrPool
+	authHandler func(string) bool
+}
+
+// ServerConfig defines all data needed for running the Server.
+type ServerConfig struct {
+	// TlsCrtFilePath specifies path to certificate file for tls connection
+	TlsCrtFilePath string
+	// TlsKeyFilePath specifies path to key file for tls connection
+	TlsKeyFilePath string
+	// Address is TCP address to listen for client connections. If empty ":0" is used.
+	Address string
+	// Address Pool enables Port AutoAssignation. If empty "10000:50000" is used.
+	PortRange string `default:"10000:50000"`
+	// AuthHandler is function validates provided auth token
+	AuthHandler func(string) bool
+	// Logger is optional logger. If nil logging is disabled.
+	Logger *log.Entry
+	// SNIAddress is optional TCP address to listen for TLS SNI connections
+	SNIAddress string
+}
+
+type serverData struct {
+	addr        string
+	portRange   string
+	tlsConfig   *tls.Config
+	listener    net.Listener
+	logger      *log.Entry
+	sniAddr     string
+	authHandler func(string) bool
+}
+
+// NewServer creates a new Server based on configuration.
+// Caller must invoke Start() on returned instance in order to start server
+func NewServer(config *ServerConfig) (*Server, error) {
+	serverData := &serverData{}
+
+	if config.Address == "" {
+		return nil, errors.New("provided empty address")
+	}
+	serverData.addr = config.Address
+
+	if config.PortRange == "" {
+		config.PortRange = "10000:50000"
+	}
+	serverData.portRange = config.PortRange
+
+	if config.TlsKeyFilePath == "" {
+		return nil, errors.New("provided tls key file path empty")
+	}
+	if config.TlsCrtFilePath == "" {
+		return nil, errors.New("provided tls cert file path empty")
+	}
+	tlsConfig, err := TlsServerConfig(
+		config.TlsCrtFilePath,
+		config.TlsKeyFilePath,
+		"")
+	if err != nil {
+		return nil, err
+	}
+	serverData.tlsConfig = tlsConfig
+
+	logger := config.Logger
+	if logger == nil {
+		l := log.New()
+		l.SetLevel(log.ErrorLevel)
+		logger = log.NewEntry(l)
+	}
+	serverData.logger = logger
+
+	if config.AuthHandler == nil {
+		return nil, errors.New("provided auth handler is nil")
+	}
+	serverData.authHandler = config.AuthHandler
+
+	serverData.sniAddr = config.SNIAddress
+
+	return newServer(serverData)
+}
+
+func newServer(serverData *serverData) (*Server, error) {
+	pPool := &AddrPool{}
+	err := pPool.Init(serverData.portRange)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create port range pool: %s", err)
+	}
+
+	listener, err := listener(serverData)
+	if err != nil {
+		return nil, fmt.Errorf("listener failed: %s", err)
+	}
+
+	logger := serverData.logger
```

Comment:
Same as client don't think we need theses as we did the same thing in NewClient
IMO in one place this logic is extra